pyogrio/_io.pyx

```diff
@@ -646,16 +646,28 @@ cdef get_features(
     ]
 
     field_data_view = [field_data[field_index][:] for field_index in range(n_fields)]
+    i = 0
+    while True:
+        if max_features > 0 and i == max_features:
+            break
 
-    for i in range(count):
         try:
             ogr_feature = exc_wrap_pointer(OGR_L_GetNextFeature(ogr_layer))
 
         except NullPointerError:
-            raise FeatureError(f"Could not read feature at index {i}") from None
+            # No more rows available, so stop reading
+            break
 
         except CPLE_BaseError as exc:
-            raise FeatureError(str(exc))
+            if "failed to prepare SQL" in str(exc):
+                raise ValueError(f"Invalid SQL query") from exc
+            else:
+                raise FeatureError(str(exc))
+
+        if i >= count:
```

Comment:
> Is there a way to test if this can ever occur?

Not that I can think of. There are so many ways you can filter, or have a combination of spatial filter and where eg. that it is difficult to judge if it can occur or not, and in what situation.

>This isn't something that the user can control, so raising the exception here is only helpful in terms of helping us identify a reproducible test case for it.
> 
> I think instead of an error, perhaps we can emit a warning, and then stop iteration?

I put it as an error because I personally, as a user, would rather have an error than a warning if the data being returned by a filter is wrong. At work we had a case where oracle returned too few rows in a specific edge case: having had an error instead of wrong data would have saved us a lot of trouble.

But, if you would rather have a warning: shoot and I'll change it.

Reply:
I think my main issue here relates to what we expect the user to do next if they encounter this error.  Rerunning likely won't help, nor will adjusting their `bbox` and / or `where` and / or other filters necessarily help.  

Maybe we raise the error, but we note that this is entirely unexpected and that they need to report via GH issue so that we can investigate and formulate a fix.  I'd like to not have us stub out a more complex fix here for a case that shouldn't (in theory) occur - but it really depends on whether or not the filters in GDAL will always return a count that is >= actual records, or if there is indeed a case where count is <= actual records.

```
raise FeatureError("GDAL returned more records than expected based on the count of records that may meet your combination of filters against this dataset.  Please open an issue on Github (https://github.com/geopandas/pyogrio/issues) to report encountering this error.") from None
```

This at least helps us know if it happens, and in the issue we can advise them to other temporary solutions, such as doing filtering after reading records.

Reply:
> I think my main issue here relates to what we expect the user to do next if they encounter this error. Rerunning likely won't help, nor will adjusting their `bbox` and / or `where` and / or other filters necessarily help.
> 
> Maybe we raise the error, but we note that this is entirely unexpected and that they need to report via GH issue so that we can investigate and formulate a fix. I'd like to not have us stub out a more complex fix here for a case that shouldn't (in theory) occur - but it really depends on whether or not the filters in GDAL will always return a count that is >= actual records, or if there is indeed a case where count is <= actual records.
> 
> ```
> raise FeatureError("GDAL returned more records than expected based on the count of records that may meet your combination of filters against this dataset.  Please open an issue on Github (https://github.com/geopandas/pyogrio/issues) to report encountering this error.") from None
> ```
> 
> This at least helps us know if it happens, and in the issue we can advise them to other temporary solutions, such as doing filtering after reading records.

Good point. I agree the only suitable work around would be not to filter using pyogrio but doing the filtering afterwards... till the behaviour is investigated and fixed somehow. Hence, making the error clearer and more explicit sounds like a good idea...